### DISCUSSION TITLE
rke2 worker node not getting IP

### DISCUSSION THREAD

**Discussion by bpopovich44** [rancher/rke2#4782] [General]
I am running RKE2 in a HA environment.  Three control-plane nodes and three worker nodes.  I needed to patch the six servers so decided to run rke2 upgrade manually from the following link:  https://docs.rke2.io/upgrade/manual_upgrade
The upgrade seemed to go fine with no errors.  After everything was back up, my cluster was not accessible from outside the cluster via kubectl or rancher.  I was getting the below error from the API so decided to restore from backup.  I restored to backup and the cluster came back up(after backup, the kubernetes version stayed on latest version).  Now though, one of my worker nodes will not get an IP so will not connect.  Will you please point me in the right direction to troubleshoot.  I also have the worker node cordoned off so no pods will be assigned to it.

**Output from worker node:**
```console
kubectl get pods -A --field-selector spec.nodeName=rke02 -o wide
NAMESPACE                  NAME                                                READY   STATUS                  RESTARTS          AGE    IP              NODE    NOMINATED NODE   READINESS GATES
cattle-monitoring-system   pushprox-kube-proxy-client-p6f8n                    1/1     Running                 13 (32h ago)      279d   192.168.121.8   rke02   <none>           <none>
cattle-monitoring-system   rancher-monitoring-prometheus-node-exporter-fnbgz   1/1     Running                 10169 (31h ago)   279d   192.168.121.8   rke02   <none>           <none>
default                    overlaytest-6zmfp                                   0/1     Unknown                 10                237d   <none>          rke02   <none>           <none>
devops-sandbox             centos7-w-affinity--rke02                           0/1     Terminating             1                 33h    <none>          rke02   <none>           <none>
devops-sandbox             livenessprobe-69f48d456d-fsnn9                      0/1     Terminating             19                33h    <none>          rke02   <none>           <none>
devops-sandbox             ubuntu-w-affinity--rke02                            0/1     Terminating             1                 33h    <none>          rke02   <none>           <none>
gitlab                     general-runner-gitlab-runner-74cb4d84cf-995nm       0/1     Terminating             0                 32h    <none>          rke02   <none>           <none>
kube-system                kube-proxy-rke02                                    1/1     Running                 0                 32h    192.168.121.8   rke02   <none>           <none>
kube-system                rke2-canal-d88tj                                    0/2     Init:CrashLoopBackOff   18 (31h ago)      34h    192.168.121.8   rke02   <none>           <none>
kube-system                rke2-ingress-nginx-controller-jpjqm                 0/1     Unknown                 1                 33h    <none>          rke02   <none>           <none>
qemu                       qemu-arm64-w-affinity-toleration--rke02             0/1     Terminating             3                 33h    <none>          rke02   <none>           <none>
```

**Error from api before restore from backup:**
```
Warning  Unhealthy  34m   kubelet  Liveness probe failed: Error from server (InternalError): an error on the server ("[+]ping ok\n[+]log ok\n[-]etcd failed: reason withheld\n[+]poststarthook/start-kube-apiserver-admission-initializer ok\n[+]poststarthook/generic-apiserver-start-informers ok\n[+]poststarthook/priority-and-fairness-config-consumer ok\n[+]poststarthook/priority-and-fairness-filter ok\n[+]poststarthook/storage-object-count-tracker-hook ok\n[+]poststarthook/start-apiextensions-informers ok\n[+]poststarthook/start-apiextensions-controllers ok\n[+]poststarthook/crd-informer-synced ok\n[+]poststarthook/bootstrap-controller ok\n[+]poststarthook/rbac/bootstrap-roles ok\n[+]poststarthook/scheduling/bootstrap-system-priority-classes ok\n[+]poststarthook/priority-and-fairness-config-producer ok\n[+]poststarthook/start-cluster-authentication-info-controller ok\n[+]poststarthook/aggregator-reload-proxy-client-cert ok\n[+]poststarthook/start-kube-aggregator-informers ok\n[+]poststarthook/apiservice-registration-controller ok\n[+]poststarthook/apiservice-status-available-controller ok\n[+]poststarthook/kube-apiserver-autoregistration ok\n[+]autoregister-completion ok\n[+]poststarthook/apiservice-openapi-controller ok\n[+]poststarthook/apiservice-openapiv3-controller ok\nlivez check failed") has prevented the request from succeeding
  Warning  Unhealthy  34m   kubelet  Readiness probe failed: Error from server (InternalError): an error on the server ("[+]ping ok\n[+]log ok\n[-]etcd failed: reason withheld\n[+]etcd-readiness ok\n[+]informer-sync ok\n[+]poststarthook/start-kube-apiserver-admission-initializer ok\n[+]poststarthook/generic-apiserver-start-informers ok\n[+]poststarthook/priority-and-fairness-config-consumer ok\n[+]poststarthook/priority-and-fairness-filter ok\n[+]poststarthook/storage-object-count-tracker-hook ok\n[+]poststarthook/start-apiextensions-informers ok\n[+]poststarthook/start-apiextensions-controllers ok\n[+]poststarthook/crd-informer-synced ok\n[+]poststarthook/bootstrap-controller ok\n[+]poststarthook/rbac/bootstrap-roles ok\n[+]poststarthook/scheduling/bootstrap-system-priority-classes ok\n[+]poststarthook/priority-and-fairness-config-producer ok\n[+]poststarthook/start-cluster-authentication-info-controller ok\n[+]poststarthook/aggregator-reload-proxy-client-cert ok\n[+]poststarthook/start-kube-aggregator-informers ok\n[+]poststarthook/apiservice-registration-controller ok\n[+]poststarthook/apiservice-status-available-controller ok\n[+]poststarthook/kube-apiserver-autoregistration ok\n[+]autoregister-completion ok\n[+]poststarthook/apiservice-openapi-controller ok\n[+]poststarthook/apiservice-openapiv3-controller ok\n[+]shutdown ok\nreadyz check failed") has prevented the request from succeedin
```

**Comment by bpopovich44** [comment]
When I look at the logs for the three kube-system pods,( kube-proxy, rke2-canal, rke2-ingress-nginx-controller), I get the same error:
Error from server: Get "https://192.168.121.8:10250/containerLogs/kube-system/rke2-ingress-nginx-controller-jpjqm/rke2-ingress-nginx-controller": proxy error from 127.0.0.1:9345 while dialing 192.168.121.8:10250, code 503: 503 Service Unavailable


There are no events when I describe the pods

**Comment by brandond** [comment]
> I am running RKE2 in a HA environment. Three control-plane nodes and three worker nodes. I needed to patch the six servers so decided to run rke2 upgrade manually

You didn't open this as an issue, so we don't have all the info that we would normally ask for via the issue template, so I have to ask - what versions did you upgrade from, and to?

>  I restored to backup and the cluster came back up

What kind of backup did you restore from? Did you restore the entire OS, or just restore an etcd snapshot to the datastore?

What versions of RKE2 are all of the nodes currently running?

**Reply by bpopovich44** [reply]
Thank you Brandon, I apologize I didn't open in the correct template.
Before the upgrade, kubernetes control-plane nodes were running v1.24.11+rke2r1.  The worker nodes were running v1.24.13+rke2r1.
Running the command **curl -sfL https://get.rke2.io | sh -** on the control-plane nodes, and **curl -sfL https://get.rke2.io | INSTALL_RKE2_TYPE="agent" sh -** on the worker nodes, everything is now running v1.25.13+rke2r1.  Just to reiterate, I did not see any errors after the upgrades.

I restored from etcd snapshot to the datastore.  The snapshot was taken the day before automatically by rke2.  


**kubectl get nodes -o wide**
```console
NAME       STATUS                        ROLES                       AGE    VERSION           INTERNAL-IP     EXTERNAL-IP   OS-IMAGE                               KERNEL-VERSION                 CONTAINER-RUNTIME
rke00      Ready                         <none>                      252d   v1.25.13+rke2r1   192.168.xxx.xxx   <none>        Rocky Linux 8.8 (Green Obsidian)       4.18.0-477.21.1.el8_8.x86_64   containerd://1.7.3-k3s1
rke01      Ready                         <none>                      253d   v1.25.13+rke2r1   192.168.xxx.xxx   <none>        Rocky Linux 8.8 (Green Obsidian)       4.18.0-477.21.1.el8_8.x86_64   containerd://1.7.3-k3s1
rke02      Ready,SchedulingDisabled   <none>                      279d   v1.25.13+rke2r1   192.168.xxx.xxx   <none>        Rocky Linux 8.8 (Green Obsidian)       4.18.0-477.21.1.el8_8.x86_64   containerd://1.7.3-k3s1
rke2a-ha   Ready                         control-plane,etcd,master   332d   v1.25.13+rke2r1   192.168.xxx.xxx   <none>        Red Hat Enterprise Linux 8.8 (Ootpa)   4.18.0-477.21.1.el8_8.x86_64   containerd://1.7.3-k3s1
rke2b-ha   Ready                         control-plane,etcd,master   332d   v1.25.13+rke2r1   192.168.xxx.xxx   <none>        Red Hat Enterprise Linux 8.8 (Ootpa)   4.18.0-477.21.1.el8_8.x86_64   containerd://1.7.3-k3s1
rke2c-ha   Ready                         control-plane,etcd,master   332d   v1.25.13+rke2r1   192.168.xxx.xxx   <none>        Red Hat Enterprise Linux 8.8 (Ootpa)   4.18.0-477.21.1.el8_8.x86_64   containerd://1.7.3-k3s1
```

**kubectl version**
```console
Client Version: version.Info{Major:"1", Minor:"23", GitVersion:"v1.23.1", GitCommit:"86ec240af8cbd1b60bcc4c03c20da9b98005b92e", GitTreeState:"clean", BuildDate:"2021-12-16T11:41:01Z", GoVersion:"go1.17.5", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"25", GitVersion:"v1.25.13+rke2r1", GitCommit:"5244794d27b4cc68290bc496b00e248857ac8b47", GitTreeState:"clean", BuildDate:"2023-08-24T04:24:59Z", GoVersion:"go1.20.7 X:boringcrypto", Compiler:"gc", Platform:"linux/amd64"}
WARNING: version difference between client (1.23) and server (1.25) exceeds the supported minor version skew of +/-1
```

**Reply by bpopovich44** [reply]
I do see that warning now with the skew difference.  Not sure if it is because I restored to a snapshot that had a earlier version of rke2

**Reply by bpopovich44** [reply]
I also wanted to note.  I've had issues with pods being run on the control-plane nodes when they should be on the worker nodes so I put a taint on the control-plane nodes:
**kubectl taint node rke2a-ha node-role.kubernetes.io/control-plane:NoSchedule**

Having this in place, and after the upgrade, the kube-proxys would not land on the control-plane nodes.  After I removed the taint, the pods started

**Reply by bpopovich44** [reply]
Also, after the upgrade/patches.  It was a different worker node that wouldn't connect.  So I rebooted the servers hoping it would realign things.  After the reboot, it was this current node that wouldn't connect.  At this point, and being production, I cordoned this node to keep prod going.  Point being, it seemed to not want to connect a third node.  Not sure if this is a coincidence or not but want to give as much info as possible

**Reply by brandond** [reply]
kube-proxy is deployed as a static pod, which doesn't go through scheduling, and therefore shouldn't be affected by taints, so I'm not sure what's going on there. You don't by any chance have your nodes configured with `disable-kube-proxy: true`, do you?

Can you clarify what you mean by the worker "not getting an IP"? It appears to have an IP in your `kubectl get nodes` output.

The errors you posted earlier indicate that the apiserver's etcd health check was failing, but you were still able to interact with the cluster so I'm not sure what the root cause of that was.

What are the **current** problems with your nodes?

**Reply by bpopovich44** [reply]
No, I do not have the nodes configured with **disable-kube-proxy: true**

I'm sorry, after the update I was able to access the cluster if I was on one of the control-plane nodes with the kubeconfig local.yaml in place, but if I was on my local machine with the same local.yaml I would get an error saying 
**The connection to the server 127.0.0.1:6443 was refused - did you specify the right host or port?** and Rancher was not available from the url.  That was my initial issue before I used etcd backup.

Currently my issue is, I didn't mean to say the node is not getting an IP, but not all the pods inside are getting it from the CNI I believe.  They say **none** under IP

```console
kubectl get pods -A --field-selector spec.nodeName=rke02 -o wide
NAMESPACE                  NAME                                                READY   STATUS                  RESTARTS          AGE    IP              NODE    NOMINATED NODE   READINESS GATES
cattle-monitoring-system   pushprox-kube-proxy-client-p6f8n                    1/1     Running                 13 (39h ago)      279d   192.168.xxx.xxx   rke02  none          none
cattle-monitoring-system   rancher-monitoring-prometheus-node-exporter-fnbgz   1/1     Running                 10169 (39h ago)   279d   192.168.xxx.xxx   rke02   none           none
default                    overlaytest-6zmfp                                   0/1     Unknown                 10                238d   **none**          rke02   none           none
devops-sandbox             centos7-w-affinity--rke02                           0/1     Terminating             1                 40h    **none**          rke02   none           none
devops-sandbox             livenessprobe-69f48d456d-fsnn9                      0/1     Terminating             19                40h    **none**          rke02   none           none
devops-sandbox             ubuntu-w-affinity--rke02                            0/1     Terminating             1                 40h    **none**          rke02  none           none
gitlab                     general-runner-gitlab-runner-74cb4d84cf-995nm       0/1     Terminating             0                 40h    **none**          rke02   none           none
kube-system                kube-proxy-rke02                                    1/1     Running                 0                 40h    192.168.xxx.xxx   rke02   none           none
kube-system                rke2-canal-d88tj                                    0/2     Init:CrashLoopBackOff   18 (39h ago)      41h    192.168.xxx.xxx   rke02   none           none
kube-system                rke2-ingress-nginx-controller-jpjqm                 0/1     Unknown                 1                 40h    **none**          rke02   none          none
qemu                       qemu-arm64-w-affinity-toleration--rke02             0/1     Terminating             3                 40h    **none**          rke02   none           none
```

**Reply by brandond** [reply]
Well yeah, the CNI pod (`rke2-canal-d88tj `) is in CrashLoopBackoff, and without the CNI none of the other pods are going to work either. I'd probably check out the logs for that pod and see why the init container is crashing. You may just need to delete it and let the daemonset redeploy a new pod.

**Comment by bpopovich44** [comment]
Hi Brandon,
I deleted rke2-canal-d88tj and the pod came back up as well as the rest of the pods.  I then again upgraded the rke2 and everything went fine the second time around.  I did delete the taint from the control-plane nodes before the second upgrade, so not sure if that made a difference or not.
So, all looks good.

Thank you!
